### PR TITLE
KIN-340 Add the functionality to send product email and update the email table in the db

### PR DESCRIPTION
**Pull request by Cristianpan**
[KIN-340](https://loopcrack.atlassian.net/browse/KIN-340) Medium [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=loopcrack-org&utm_campaign=badge_info)

### Problem

The modal form of the product page must be stored when submitted by the user, as well as it is needed to store the dates of the forms submitted by the users.

### Solution

Implement the logic to send and store the product email and update the database to add the field to store the sending dates of the mailing forms. 

### Requirements checklist

- [x] The product form must be stored as an email and sent.
- [x] The product form must be validated before submit and display the error messages.
- [x] If the product e-mail form is submitted, a success message should be displayed, otherwise, an error message should be displayed.
- [x] If the data is incomplete, it must display the errors on the form. 
- [x] The emails table of the database must have a date column for date storage.
- [x] The functionality for the technical support and contact form of the public view must have integrated the storage of the email date

### Testing
Note: Before testing this feature, is necessary that you make in the console `php spark migrate:refresh` and `php park db:seed MainTestSeeder` so you update your db.

- [x] Go to the product page and send the product form with empty data and you should see a modal with a error message. 
- [x] Go to the product page and send the  product form with all data and you sould see a modal with a success message. 
- [x] Verify that your db have the field for the date. 
- [x] Check that after sending the email, the date of sending is recorded in the database.

## Demo

[KIN-340]: https://loopcrack.atlassian.net/browse/KIN-340?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by CarlosMay7 — COMMENTED** [review]
It works fine overall but I have a question. the date must be the same as the time it was sent, right? Because it didn't work for me

image
image

**Comment by CarlosMay7** [comment]
Also, change the pull request name to KIN-340...

**Comment by Cristianpan** [comment]
> It works fine overall but I have a question. the date must be the same as the time it was sent, right? Because it didn't work for me
> 
> image image

I had not considered it but have already made the change.

**Review by CarlosMay7 — APPROVED** [review]
Nice work! Thank you

**Review by FernandoJoachin — APPROVED** [review]

**Review by CarlosJCD — APPROVED** [review]
Nice work, just check out the comment i left you

**Review by Mauriciocr207 — APPROVED** [review]

File: app/Controllers/CtrlEmail.php [hunk 1]
```diff
@@ -67,6 +70,83 @@ public function sendContactEmail(): RedirectResponse
...
+            $response = [
+                'title'   => 'Envío exitoso',
+                'message' => 'Se ha enviado correctamente',
+                'type'    => 'success',
+            ];
+
+            return redirect('producto')->with('response', $response);
+        } catch (InvalidInputException $th) {
+            $response = [
+                'title'   => 'Ops! No se ha realizado el envio del email',
```
Comment (CheloChidoone) [inline app/Controllers/CtrlEmail.php:133]: In this message, it should be `Oops` not `Ops`

File: app/Controllers/CtrlEmail.php [hunk 2]
```diff
@@ -67,6 +70,83 @@ public function sendContactEmail(): RedirectResponse
...
+            $data = [
+                'emailTypeId'   => PRODUCT_EMAIL_ID,
+                'inquirerName'  => $productEmailData['inquirer-name'],
+                'inquirerEmail' => $productEmailData['inquirer-email'],
+                'emailContent'  => EmailSender::getEmailBody($formData, 'mailDetail'),
+            ];
+            (new EmailModel())->insert($data);
+            $response = [
+                'title'   => 'Envío exitoso',
+                'message' => 'Se ha enviado correctamente',
```
Comment (CheloChidoone) [inline app/Controllers/CtrlEmail.php:126]: This message is too generic, take for example the one Luis added for the `Soporte Tecnico` functionality and write a similar message here

File: app/Controllers/CtrlEmail.php [hunk 3]
```diff
@@ -106,13 +186,13 @@ public function sendSupportEmail(): RedirectResponse
 
         if ($successEmail) {
             $response = [
-                'title'   => 'Mensaje enviado correctamente',
-                'message' => 'El Formulario se ha enviado a nuestro equipo de soporte técnico, trataremos de ponernos en contacto con usted lo más pronto posible.',
+                'title'   => 'Envío exitoso',
+                'message' => 'Se ha enviado correctamente',
```
Comment (CheloChidoone) [inline app/Controllers/CtrlEmail.php:190]: Why did you change the messages for `soporte tecnico`?, you should return them to how they were

File: app/Controllers/CtrlEmail.php [hunk 4]
```diff
@@ -121,7 +201,7 @@ public function sendSupportEmail(): RedirectResponse
         } else {
             $response = [
                 'title'   => 'Envío fallido',
-                'message' => 'No se pudo realizar el envío del formulario',
+                'message' => 'No se pudo realizar el envío del email',
```
Comment (CheloChidoone) [inline app/Controllers/CtrlEmail.php:204]: Same, Why did you change the messages for `soporte tecnico`?, you should return them to how they were

**Review by CheloChidoone — CHANGES_REQUESTED** [review]
I think there are many things to fix @Cristianpan

1. I think it should validate the form before sending it, not show the error message and then we have to reopen it to see the errors we had made when filling it
2. When I left the form empty and the error messages appears on the form it makes the modal bigger than my screen and the bottom part doesn't show and I can't scroll on it either 
image
Also check my other comments about the response messages.

File: app/Controllers/CtrlEmail.php (hunk 3 above)
Comment (Cristianpan) [reply app/Controllers/CtrlEmail.php]: It was my mistake, I've already corrected it.

File: app/Controllers/CtrlEmail.php (hunk 4 above)
Comment (Cristianpan) [reply app/Controllers/CtrlEmail.php]: It was my mistake, I've already corrected it too.

File: app/Config/Constants.php [hunk 5]
```diff
@@ -101,6 +101,13 @@
 define('FILES_TEMP_DIRECTORY', './files/tmp/');
 define('FILES_UPLOAD_DIRECTORY', './uploads/');
 
+/**
+ * Email Types ID
+ */
+define('CONTACT_EMAIL_ID', 1);
+define('SUPPORT_EMAIL_ID', 2);
+define('PRODUCT_EMAIL_ID', 3);
```
Comment (LUGAMAFE) [inline app/Config/Constants.php:109]: Hey @Cristianpan
```suggestion
```

Let move these constants to the file `app\Controllers\CtrlEmail.php`

```php
<?php
namespace App\Controllers;

class CtrlEmail extends BaseController
{
    /**
     * Email Types ID
     */
    const CONTACT_EMAIL_ID = 1;
    const SUPPORT_EMAIL_ID = 2;
    const PRODUCT_EMAIL_ID = 3;

    // Resto del código...
}
?>```

so we can use it like this: `CtrlEmail::CONTACT_EMAIL_ID;`

i don't see those variables being used in other places

File: src/public/js/product.js [hunk 6]
```diff
@@ -175,10 +195,66 @@ modalClose.addEventListener('click', function () {
...
+
+const configValidator = {
+  container: '.modal-form',
+  validateOnFieldChanges: true,
+  button: '.modal-form__submit',
+  customValidates: {
+    intlTelInput: {
+      message: errorMap[0],
+      // eslint-disable-next-line
+      fn: function (field, container) {
```
Comment (LUGAMAFE) [inline src/public/js/product.js:242]: Remove this line and remove unused variable
```suggestion
      fn: function (field) {
```

File: src/public/js/product.js [hunk 7]
```diff
@@ -175,10 +195,66 @@ modalClose.addEventListener('click', function () {
...
+const commonCustomViewErrors = {
+  add: function (field, message, cls) {
+    let error = document.querySelector(`.${field.id}.${cls}`) || document.createElement('p');
+    error.textContent = message;
+
+    if (!document.querySelector(`.${field.id}.${cls}`)) {
+      error.classList.add(field.id, cls);
+      clonedForm.querySelector(`[for='${field.id}']`).after(error);
+    }
+  },
```
Comment (LUGAMAFE) [inline src/public/js/product.js:219]:
```suggestion
  add: function (field, message, cls) {
    let errorElement = document.querySelector(`.${field.id}.${cls}`);
    
    if (!errorElement) {
      errorElement = document.createElement('p');
      errorElement.classList.add(field.id, cls);
      clonedForm.querySelector(`[for='${field.id}']`).after(errorElement);
    }

    errorElement.textContent = message;
  },
```

**Review by LUGAMAFE — CHANGES_REQUESTED** [review]

**Review by LUGAMAFE — CHANGES_REQUESTED** [review]
Good job @Cristianpan just please correct the nits and please make the mobile view look like this:

image

**Review by CarlosJCD — APPROVED** [review]
Nice Job!, just check out @CheloChidoone requested changes.

**Review by CheloChidoone — CHANGES_REQUESTED** [review]
Good Job @Cristianpan, thanks for making my requested changes, I have one more, make sure to deactivate the `enviar` button after the first click to prevent multiple emails

CC: @CarlosJCD

**Review by CheloChidoone — APPROVED** [review]
Good Job @Cristianpan

**Review by LUGAMAFE — CHANGES_REQUESTED** [review]
Please check and remove this padding on mobile views:
image

And merge with the latest main branch please

**Review by LUGAMAFE — APPROVED** [review]
Thanks @Cristianpan

**Review by Mauriciocr207 — APPROVED** [review]

**Pull request merged**